src/common/upload.ts

```diff
@@ -56,6 +58,87 @@ export const creditSharingTagNames = {
 export const developmentUploadServiceURL = 'https://upload.ardrive.dev';
 export const defaultUploadServiceURL = 'https://upload.ardrive.io';
 
+export type UploadEmitterEvent = 'progress';
+
+export type UploadProgressEvent = {
+  chunk: Buffer;
+};
+
+export type UploadEmitterParams = {
+  onProgress?: (event: UploadProgressEvent) => void;
+};
+
+export class UploadEmitter extends EventEmitter<UploadEmitterEvent> {
+  constructor(params?: UploadEmitterParams) {
+    super();
+    if (params?.onProgress !== undefined) {
+      this.on('progress', params.onProgress);
+    }
+  }
+
+  static from(params?: UploadEmitterParams | UploadEmitter) {
+    if (params instanceof UploadEmitter) {
+      return params;
+    }
+    return new UploadEmitter(params);
+  }
+
+  // todo: create listener params type
+  on(event: 'progress', listener: (ctx: UploadProgressEvent) => void): this;
+  on(event: UploadEmitterEvent, listener: (...args: any[]) => void): this {
```

Comment:
Is there an opportunity here to be more specific about types for the listener function?